lib/internal/process/esm_loader.js

```diff
@@ -72,17 +89,11 @@ async function initializeLoader() {
   const internalEsmLoader = new ESMLoader();
 
   // Importation must be handled by internal loader to avoid poluting userland
-  const keyedExportsList = await internalEsmLoader.import(
-    customLoaders,
+  return internalEsmLoader.import(
+    specifiers,
```

Comment:
We shouldn’t reuse the logic for loaders to support the modules in `--import`, because if any of the files/packages in `--import` export a function with a name we’re looking for as a hook (`resolve`, `load`, `globalPreload`) then it’ll get parsed as a loader and not as arbitrary user code. Only files registered via `--loader` should potentially register hooks.

Reply:
this function only loads the module without looking at its exports,
only later on at `addCustomLoaders` (which is called only for `--loader`) the exports (`resolve`, `load` etc) are accessed

Reply:
When it “loads the module” is it evaluating the code?

Flags like `--import a.js --loader b.js --import c.js --loader d.js` should evaluate the loaders (b and d) first; and those are used for module resolution/loading within a and c. Is that happening currently?

We’ll still need to separate these once we move loaders off-thread, since the modules loaded by `--import` would stay on the main thread.

Reply:
`ESMLoader::import()` is safe to use in terms of logic/pollution, regardless of hooks (it's actually what is eventually used to facilitate a dynamic import). Its only side effect is an addition to module cache.

(`ESMLoader::addCustomLoaders()` is indeed the done sensitive to hook exports)

Reply:
> Flags like `--import a.js --loader b.js --import c.js --loader d.js` should evaluate the loaders (b and d) first; and those are used for module resolution/loading within a and c. Is that happening currently?

@GeoffreyBooth yes 🙂 line 61 evals loaders and then preload imports on line 70

https://github.com/nodejs/node/blob/ea1597f034f1ce0495c3c230da1cb0650c7c3dc7/lib/internal/process/esm_loader.js#L61

https://github.com/nodejs/node/blob/ea1597f034f1ce0495c3c230da1cb0650c7c3dc7/lib/internal/process/esm_loader.js#L70

Reply:
I will add a test for combining `--loader` and `--import`